### ISSUE TITLE
New rule: `UnnecessaryChainedSafeCall`

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#9622]
## Expected Behavior of the rule
Some times I find this patter in the code: `a.b?.c?.d?.e!!` and it can be replaced with `a.b!!.c.d.e`.

Even if `d` is also nullable `a.b!!.c.d!!.e` would be better.

This rule should also check for usages of `requireNotNull` or `checkNotNull`.

## Context
This is similar to `ReplaceSafeCallChainWithRun`. Both rules try to attack the same problem: unnecessary chained safe calls.

There is some relation with `UnnecessarySafeCall` but it's remote. This one find cases when the safe call can be just removed.

cc @3flex `ReplaceSafeCallChainWithRun` was porposed and implemented by you. What do you think about rename it to `UnnecessaryChainedSafeCall` and not only recommend to use `run` but also recommend an early non-null assertion over a late non-null assertion.

**Comment by rastiehaiev** [comment]
@BraisGabin I hope you don’t mind me joining the discussion. I’ve been looking into this issue and wanted to share my thoughts about the rule’s scope.

### Create new or update existing rule?

I think this should be implemented as a new rule rather than by extending or renaming `ReplaceSafeCallChainWithRun`.

Although both rules analyze safe-call chains, they solve different problems:

- `ReplaceSafeCallChainWithRun` preserves nullable behavior. If the initial receiver is null, the complete expression still evaluates to null. Its purpose is to remove redundant safe calls and reduce complexity.
- The proposed rule handles chains where null is ultimately not accepted because the result is passed to `!!`, `requireNotNull`, or `checkNotNull`. Its purpose is to move the failure closer to the nullable value whose absence violates an invariant.

The two rules may be able to share some internal safe-call-chain analysis, but I think their public behavior and intent are different enough to keep them separate. I suggest naming the new rule `AvoidLateNonNullAssertion`.

### Use-cases analysis

For chains ending with `!!`, the assertion can be moved to each value that is actually nullable.

Handling `requireNotNull` and `checkNotNull` is more complicated. When exactly one value is nullable, the intended improvement is straightforward:

```kotlin
requireNotNull(a.b?.c?.d?.e)
```

could be rewritten by the developer as:

```kotlin
requireNotNull(a.b).c.d.e
```

However, when multiple values are nullable, directly moving every check could produce difficult-to-read nested code:

```kotlin
requireNotNull(requireNotNull(a.b).c.d).e
```

It becomes even less readable when custom messages are present:

```kotlin
requireNotNull(
    requireNotNull(a.b) { "message" }.c.d
) { "message" }.e
```

The same problem applies to `checkNotNull`.

### Suggestion

I suggest that the rule should still report these cases. The rule could use one generic finding message for all supported cases:

> Move non-null checks closer to the nullable values instead of checking the result of the safe-call chain.

This message would advise the developer to move the checks earlier, and when multiple nullable values are involved, the developer could split the chain into local variables and separate checks in the way that best expresses the relevant invariants.

What do you think about this scope and the `AvoidLateNonNullAssertion` name?

**Comment by BraisGabin** [comment]
> [@BraisGabin](https://github.com/BraisGabin) I hope you don’t mind me joining the discussion. I’ve been looking into this issue and wanted to share my thoughts about the rule’s scope.

Please do! Join any discussion with your thoughts. I must say that the writing looks LLM-like. If that's the case I would prefer to read your thoughts with your words

---

My idea of this rule is not to move the check closer (that's a *really* nice side effect) is to make the code easier to read. And `?.` is the problem in this cases. That's the same reason that I see behind `ReplaceSafeCallChainWithRun`.

About multiple `requireNotNull`/`checkNotNull` I agree with you that it makes the code a mess. But, to me, a single `requireNotNull`/`checkNotNull` makes the code more difficult to read than `!!`. There are places were `requireNotNull`/`checkNotNull` is a good idea but I'm a big fan of `!!`.

A sample: `requireNotNull(Masters.ash.pokemon.pikachu)` when I read `pikachu` I already forget that this code was inside `requireNotNull`. With `Masters.ash.pokemon.pikachu!!` the check is just in the place where the null can happen.

So

> It becomes even less readable when custom messages are present

true! But only use one `requireNotNull` when there are 2 things that can be nullable doesn't feel right. You are hiding that fact.

**Comment by rastiehaiev** [comment]
> I must say that the writing looks LLM-like. If that's the case I would prefer to read your thoughts with your words

These _are_ my thoughts. I use LLM just to present it in a more correct and structured way. I do multiple iterations before I send it so it sounds more like me :) But if you prefer if I use my words, I'll do that from now on.

> My idea of this rule is not to move the check closer (that's a really nice side effect) is to make the code easier to read.

Oh I was trying to solve another problem. Let's have an example:

```
fun main() {
    class D
    class C(val d: D?)
    class B(val c: C?)
    class A(val b: B?)

    val a = A(B(null))

    println(a.b?.c?.d!!)
}
````

Putting `!!` at the end makes the problem troubleshooting hard because it is not obvious whether `b`, `c`, or `d` was null.

> but I'm a big fan of !!

I'm not tbh :) I would prefer `requireNotNull` and `checkNotNull` with a `lazyMessage`, or even this:
```
val b = a.b ?: error("b is null")
```
because it shows exactly what was `null` and may save developer's time troubleshooting the problem. In this particular case, the sequence of `?.` doesn't bother me much.

Another point: if we do the early assertion, in this particular case it will be hard to read as well:

```
println(a.b!!.c!!.d!!) // vs println(a.b?.c?.d!!)
```

**Comment by BraisGabin** [comment]
> Another point: if we do the early assertion, in this particular case it will be hard to read as well:
> 
> ```
> println(a.b!!.c!!.d!!) // vs println(a.b?.c?.d!!)
> ```

For sure, but at least the code shows the intention: the code gambles 3 times for a NPE, instead of hide 2 of those behind `?.`. Anyway, I think that we both agree in the important points. And at the end this will be its own rule (https://github.com/detekt/detekt/pull/9671#pullrequestreview-5061372599)
